Migrate easy q(u)int/bits stuff to torch/headeronly

janeyx99 opened the pull request:
Straightup copy pasta. Keeps APIs in c10 and reexposes them to torch::headeronly.

It is arguable that we should just get rid of some of these unused dtypes but that is outside the scope of this PR, which is meant to build up to ScalarType moving to headeronly. 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #159302

janeyx99 committed:
Migrate easy q(u)int/bits stuff to torch/headeronly

[ghstack-poisoned]

desertfire reviewed (commented):
Shall we have some aoti_abi_check  tests?

janeyx99 committed:
Update on "Migrate easy q(u)int/bits stuff to torch/headeronly"

Straightup copy pasta. Keeps APIs in c10 and reexposes them to torch::headeronly.

It is arguable that we should just get rid of some of these unused dtypes but that is outside the scope of this PR, which is meant to build up to ScalarType moving to headeronly. 




[ghstack-poisoned]

janeyx99 committed:
Update on "Migrate easy q(u)int/bits stuff to torch/headeronly"

Straightup copy pasta. Keeps APIs in c10 and reexposes them to torch::headeronly.

It is arguable that we should just get rid of some of these unused dtypes but that is outside the scope of this PR, which is meant to build up to ScalarType moving to headeronly. 




[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge

albanD reviewed (approved):
Thx!

janeyx99 commented:
@pytorchbot merge -r

janeyx99 committed:
Update

[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge
